### DISCUSSION TITLE
Slower than automatic1111, help me configure it better

### DISCUSSION THREAD

**Discussion by gruevy** [vladmandic/sdnext#288] [General]
I tried this fork because I thought I used the new TensorRT thing that Nvidia put out but it turns out it runs slower, not faster, than automatic1111 main. I use --opt-sdp-attention instead of xformers because it's easier and the performance is about the same, and it looks like it works in both repos. I'm still trying to get anywhere close to the 40 it/s that some people are getting with their 4090s and having no luck. Here are my benchmarks: 

This branch:

  2023-04-20 08:37:12.965865
	  
  7.67 / 11.89 / 15.19
	  
  updated:2023-04-20 hash:f2c3978a url:https://github.com/anapnoe/stable-diffusion-webui-ux.git/tree/master
	  
  arch:AMD64 cpu:AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD system:Windows release:Windows-10-10.0.22621-SP0 python:3.10.7
	  
  torch:1.13.1+cu117 autocast half xformers:unavailable accelerate:0.12.0 transformers:4.25.1
	  
  device:NVIDIA GeForce RTX 4090 (1) (compute_37) (8, 9) cuda:11.7 cudnn:8500 24GB
	  
  Doggettx none
	  
  v1-5-pruned.safetensors [5929c1736f]

automatic1111 main: 

  2023-04-20 09:11:57.833105
	  
  10.06 / 15.73 / 17.67
	  
  updated:2023-03-29 hash:22bcc7be url:https://github.com/AUTOMATIC1111/stable-diffusion-webui/tree/master
	  
  arch:AMD64 cpu:AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD system:Windows release:Windows-10-10.0.22621-SP0 python:3.10.7
	  
  torch:1.13.1+cu117 autocast half xformers:unavailable accelerate:0.12.0 transformers:4.25.1
	  
  device:NVIDIA GeForce RTX 4090 (1) (compute_37) (8, 9) cuda:11.7 cudnn:8800 24GB
	  
  Doggettx none
	  
  v1-5-pruned-emaonly.safetensors [6ce0161689]

**Comment by vladmandic** [comment]
SDP is only available on torch 2.0 and you have torch 1.13.
uninstall torch and let installer install correct version.
same for other system libs - your accelerate is ancient - this is not what installer here installs.

as a result, you can clearly see you're not using SDP, you're using `Doggettx` - its printed in the output (and on the console when you start server)

**Reply by gruevy** [reply]
Thanks. I'll see if I can figure out how to uninstall those and run the installer fresh.

**Reply by gruevy** [reply]
So I uninstalled Torch (I'm pretty sure) 

PS D:\stable diffusion\stable-diffusion-webui-ux> pip3 uninstall torch
Found existing installation: torch 1.13.1
Uninstalling torch-1.13.1:
  Would remove:
    c:\users\ryan\appdata\local\programs\python\python310\lib\site-packages\functorch\*
    c:\users\ryan\appdata\local\programs\python\python310\lib\site-packages\torch-1.13.1.dist-info\*
    c:\users\ryan\appdata\local\programs\python\python310\lib\site-packages\torch\*
    c:\users\ryan\appdata\local\programs\python\python310\lib\site-packages\torchgen\*
    c:\users\ryan\appdata\local\programs\python\python310\scripts\convert-caffe2-to-onnx.exe
    c:\users\ryan\appdata\local\programs\python\python310\scripts\convert-onnx-to-caffe2.exe
    c:\users\ryan\appdata\local\programs\python\python310\scripts\torchrun.exe
Proceed (Y/n)? y
  Successfully uninstalled torch-1.13.1
PS D:\stable diffusion\stable-diffusion-webui-ux> cd ..
PS D:\stable diffusion> pip3 uninstall torch
WARNING: Skipping torch as it is not installed.

I also deleted the stable-diffusion-webui-ux and did a new git clone to reinstall it. After running the webui-user.bat and doing another benchmark I got this: 

 2023-04-20 09:46:29.229212
	
8.29 / 12.58 / 16.2
	
updated:2023-04-20 hash:f2c3978a url:https://github.com/anapnoe/stable-diffusion-webui-ux.git/tree/master
	
arch:AMD64 cpu:AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD system:Windows release:Windows-10-10.0.22621-SP0 python:3.10.7
	
torch:1.13.1+cu117 autocast half xformers:unavailable accelerate:0.12.0 transformers:4.25.1
	
device:NVIDIA GeForce RTX 4090 (1) (compute_37) (8, 9) cuda:11.7 cudnn:8500 24GB
	
Doggettx none
	
v1-5-pruned.safetensors [5929c1736f]

I know you didn't sign up to do tech support for a hundred feckless users but if you could point me in the right direction I'd appreciate it. Also, how do I update my "accelerate"? Not sure what that refers to.

**Reply by vladmandic** [reply]
But this is a completely different repo - see URL in the log. And that repo installs torch 1.13

**Reply by gruevy** [reply]
Ah, got it. I dunno how I did that. Probably had multiple tabs open and copied the wrong one. Thanks again

**Reply by gruevy** [reply]
Update: Thanks a ton. Check this shit out

2023-04-20 10:38:28.906840
	
14.96 / 24.95 / 33.64
	
updated:2023-04-20 hash:752b91d3 url:https://github.com/vladmandic/automatic/tree/master
	
arch:AMD64 cpu:AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD system:Windows release:Windows-10-10.0.22621-SP0 python:3.10.7
	
torch:2.0.0+cu118 Autocast half xformers:unavailable accelerate:0.18.0 transformers:4.26.1
	
device:NVIDIA GeForce RTX 4090 (1) (compute_37) (8, 9) cuda:11.8 cudnn:8700 24GB
	
sdp none
	
v1-5-pruned.safetensors [5929c1736f]

**Reply by vladmandic** [reply]
so this is pretty much double your original performance.
you can retry using extra steps and extensive level and see how far that gets you?

**Reply by gruevy** [reply]
Here's a test with extra steps and extensive. I'm pretty pleased.

2023-04-20 10:58:54.010766
	
16.19 / 31.81 / 41.67 / 40.72 / 40.76
	
updated:2023-04-20 hash:752b91d3 url:https://github.com/vladmandic/automatic/tree/master
	
arch:AMD64 cpu:AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD system:Windows release:Windows-10-10.0.22621-SP0 python:3.10.7
	
torch:2.0.0+cu118 Autocast half xformers:unavailable accelerate:0.18.0 transformers:4.26.1
	
device:NVIDIA GeForce RTX 4090 (1) (compute_37) (8, 9) cuda:11.8 cudnn:8700 24GB
	
sdp none
	
lyriel_v15.safetensors [4d91c4c217]

**Reply by vladmandic** [reply]
now you're getting to expected levels :)

**Reply by gruevy** [reply]
Feels good man :D

**Reply by Squeezitgirdle** [reply]
I'm also on a 4090, would you mind sharing your settings with me?  Mine runs fine but I've been running into a lot of memory issues or where it starts generating then just stops and never finishes or gives any error.

I probably have something configured incorrectly.

**Comment by Michoko92** [comment]
Hi,

I'd love to adopt this repo, but despite all my efforts, I can't obtain the same speed as Auto1111's original repo.

I updated Auto1111 some time ago to latest torch + xformers, and am getting around 5.2 it/s with xformers (which for me performs better than SDP on my RTX 2060 6 GB VRAM card). So I copied my xformers repo to Vlad's one, activated xformers in the settings, then restarted. Vlad's System Info tells me this:

```
Torch
2.0.0+cu118 Autocast  half
GPU
device: NVIDIA GeForce RTX 2060 (1) (compute_37) (7, 5)
cuda: 11.8
cudnn: 8700
Memory optimization
medvram
0.2/0.9s
Cross-attention
xformers
Libs
xformers: 0.0.17
accelerate: 0.18.0
transformers: 4.26.1
```

All seems good. I compared both Auto1111 and Vlad1111 settings one by one, and they are identical. And despite this, I barely get 4 it/s with Vlad's (vs 5.2 it/s with Auto).

I tried SDP in Vlad's, as I saw it was recommended in several threads, but it's not better for me, like in Auto1111's one. I'm a bit out of options, as everything seems to be configured the same in both repos. Would you have any suggestion, please?

**Reply by AlUlkesh** [reply]
For speed comparisons I think you get better results if you use the benchmark from the `System Info` tab/extension. Can you try that out and post the results?

**Reply by Michoko92** [reply]
Thank you for your reply. Here are the results:

Auto1111
![Auto](https://user-images.githubusercontent.com/3066173/233987148-f8192596-8a2d-4963-ac87-48ae71f5ffa7.jpg)


Vlad1111
![Vlad](https://user-images.githubusercontent.com/3066173/233987237-b1005c1b-26f6-4af0-be00-385d946017c8.jpg)

If I understand correctly, Auto is faster on smaller batch sizes, while Vlad's is faster on bigger ones?

**Reply by AlUlkesh** [reply]
You're using different models. Try the same one for comparisons.

**Reply by Michoko92** [reply]
OK here they are (and thank you for your time :) )

Auto
![Auto](https://user-images.githubusercontent.com/3066173/234007891-f0d67fcd-a8a5-4735-a015-08f30f686266.jpg)

Vlad
![Vlad](https://user-images.githubusercontent.com/3066173/234007948-35fe2917-fd18-453a-a55b-e8a5cb8d31e7.jpg)

**Reply by Michoko92** [reply]
Hello,

I suppose nobody knows why I'm experiencing this? :) If I'm not mistaken, the benchmark above shows a significant performance drop with Vlad's repo when generating single images (from 4.41 it/s down to 3.63 it/s). Could it be related to the "accelerate" or "transformers" libraries, as it seems they are the only difference between the two? Thank you.

**Reply by Gazzoo-byte** [reply]
Could potentially be the difference in xformers - try using --opt-sdp-attention in place of --xformers in A1111 and the following settings in Vlad (might need to restart the latter from scratch for the setting change to take effect? Not 100% certain)

![image](https://user-images.githubusercontent.com/73721238/234599838-7f83d90a-3774-4005-ac25-f6b5782d0066.png)

**Reply by Michoko92** [reply]
Thank you for your reply and your benchmark, it's much appreciated. Unfortunately, xformers have always performed better on my RTX 2060, that's why I wanted to use this option instead of SDP. I was wondering if Vlad's repo handled xformers differently, as it seems I use the same libraries on both repos.

**Reply by Gazzoo-byte** [reply]
Also, out of curiosity, could you try the benchmark tool without --medvram ? I wonder if there's something different there between A1x4 and Vlad, I don't imagine it will cause any major shift, but it may be worth investigating just in case

**Reply by Michoko92** [reply]
Good suggestion, thanks. Here are the results. Auto is still a bit better with single images, while Vlad is definitely faster with bigger batches. Seems like VRAM optimizations might work differently in both repos.

Auto
![Auto](https://user-images.githubusercontent.com/3066173/234612549-8b7f8ee4-3c3d-409e-a021-bec9b659fd08.jpg)

Vlad
![Vlad](https://user-images.githubusercontent.com/3066173/234612633-ca5fd80f-f90b-48c9-b22d-00c654c5443e.jpg)

**Reply by Gazzoo-byte** [reply]
May be a long shot, but it seems there's a newer version of xformers available now, might be worth trying it to see how you fare? pip install xformers==0.0.18 (or pip install xformers==0.0.19.dev524 if you want to try the bleeding edge, released 2hrs ago), could be that there's some negative interaction with the newer accelerate/transformers version that could have been improved, would need to fully review the xformers changelogs to know for sure that said I can't see a link to the changelogs on pypi

**Reply by Michoko92** [reply]
Thank you, might be worth a try, I'll just have to dive again into the instructions on how to compile them :)

**Reply by Michoko92** [reply]
I installed v0.0.18, but the launch script downgrades back to v0.0.17. I checked the requirements.txt file, but can't see a mention to the xformers version. Is there a way to prevent the launcher from downgrading xformers, please?

**Reply by vladmandic** [reply]
@Michoko92 i've provided answer in your separate thread.

**Reply by Michoko92** [reply]
Oh thank you Vlad, I'm not the creator of this thread, but I appreciate the help :)

**Reply by Gazzoo-byte** [reply]
@Michoko92 https://github.com/vladmandic/automatic/discussions/524 was a thread I made as I ran into the same issue lol, though the solution I arrived at was much less elegant - requirements.txt gets checked after xformers gets rolled back to 0.0.17, so I added xformers==0.0.18 to requirements.txt temporarily. Leads to xformers being downgraded then upgraded again so vlad's advice in the thread is vastly better

**Reply by Michoko92** [reply]
OK made a test with 0.0.18, and unfortunately, results are actually a bit worse than with 0.0.17 (with medvram)

![Vlad](https://user-images.githubusercontent.com/3066173/234641466-285e5fb6-1226-4383-b805-c69b5f10245f.jpg)

That's frustrating, because I'm really hyped about this new repo. But for now, Auto1111 is still performing better on single image generation with VRAM optimizations. Seems I'll have to deal with it.

**Comment by Gazzoo-byte** [comment]
I decided to approach this from a different angle by doing a bunch of generations and noting the "Time taken"
![image](https://user-images.githubusercontent.com/73721238/234593816-7d3d7061-be4f-4f46-8ce0-1132f0a11621.png)

Both used fresh pulls of https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/22bcc7be428c94e9408f589966c2040187245d81 and  https://github.com/vladmandic/automatic/commit/da35bfb7467a171548b2cca99bf9d700f62c9bc2 respectively.

Settings for generation used were DPM++ 2M Karras, 512x512, 28 steps, both using torch: 2.1.0.dev20230424+cu118 with --opt-sdp-attention, live previews disabled, lastly same model and VAE in both instances

Only thing that notibly altered the time taken was when the token length went >75 - however the difference was still within margin of error for both
